pkg/services/object/get/ec.go

```diff
@@ -355,6 +361,506 @@ func (s *Service) getECPartFromNode(ctx context.Context, cnr cid.ID, parent oid.
+	//  is only possible in bidirectional streams). This way, background buffering of
+	//  data from all streams will start immediately. Explore gRPC abilities of
+	//  buffer control for parallelism potential.
+
+	failedPartIdx, failedPartWritten, written, err := s.copyECPartsRanges(stageCtx, dst, *localNodeKey, cnr, parent, sTok, bTok, rule, ruleIdx, sortedNodes,
+		fullPartLen, firstPartIdx, firstPartOff, lastPartIdx, lastPartTo)
+	if err == nil || errors.Is(err, apistatus.ErrObjectAlreadyRemoved) || errors.Is(err, apistatus.ErrObjectOutOfRange) ||
+		errors.Is(err, errStreamFailure) || errors.Is(err, ctx.Err()) {
+		return written, err
+	}
+
+	s.log.Info("copying ranges of EC data parts failed, trying recovery...",
+		zap.Stringer("container", cnr), zap.Stringer("object", parent), zap.Stringer("rule", rule), zap.Uint64("off", off),
+		zap.Uint64("len", ln), zap.Int("failedPartIdx", failedPartIdx), zap.Uint64("failedPartWritten", failedPartWritten),
+		zap.Uint64("written", written), zap.Error(err))
+
+	if failedPartIdx != firstPartIdx {
+		firstPartOff = 0
+	}
+	firstPartIdx = failedPartIdx
+	firstPartOff += failedPartWritten
+
+	// Recover failed part. To do this, we need to fetch all other data parts (some
+	// of them may be requested too) and at least 1 parity part. Also, reserve 10%
+	// of the remaining context for maths.
+	//
+	// Note: we reread the parts that were successfully read at the previous stage. This is done to favor best case
+	// (as most expected), so as not to retain too much data in memory. Anyway, recovery requires obtaining
+	// ranges that were not originally requested.
+	if deadlineSet {
+		if stageTimeout = time.Until(deadline) * 9 / 10; stageTimeout <= 0 {
+			return written, context.DeadlineExceeded
+		}
+	} // else keep the same
+	stageCtx, cancel = context.WithTimeout(ctx, stageTimeout)
+	defer cancel()
+
+	parts, err := s.getRecoveryECPartRanges(stageCtx, *localNodeKey, cnr, parent, sTok, bTok, rule, ruleIdx, sortedNodes, fullPartLen, failedPartIdx)
+	if err != nil {
+		return written, fmt.Errorf("get recovery EC part ranges: %w", err)
+	}
+
+	if err := iec.DecodeRange(rule, firstPartIdx, lastPartIdx, parts); err != nil { // should never happen if we count parts correctly
+		return written, fmt.Errorf("recover EC part range from=%d to=%d using single parity part: %w", firstPartIdx, lastPartIdx, err)
+	}
+
+	for partIdx := firstPartIdx; partIdx <= lastPartIdx; partIdx++ {
+		from, to := uint64(0), fullPartLen
+		if partIdx == firstPartIdx {
+			from = firstPartOff
+		}
+		if partIdx == lastPartIdx {
+			to = lastPartTo
+		}
+
+		if err := dst.WriteChunk(parts[partIdx][from:to]); err != nil {
+			return written, fmt.Errorf("%w: write [%d:%d] range of EC part #%d: %w", errStreamFailure, from, to, partIdx, err)
+		}
+
+		written += to - from
+	}
+
+	return written, nil
+}
+
+func (s *Service) getECParentPayloadLen(ctx context.Context, localNodeKey ecdsa.PrivateKey, cnr cid.ID, id oid.ID, sTok *session.Object,
```

Comment:
it's worth noting that method of obtaining parent's payload len will change to GET first part in 17ca59f4af6015df6ad859bf3eca65e99dd62f2a

i'd change it right here, but for now I'll leave the original version